smartapps/curb/curb-energy-monitor.src/curb-energy-monitor.groovy

```diff
@@ -0,0 +1,477 @@
+            }
+        }
+    }
+}
+
+def oauthInitUrl() {
+    state.oauthInitState = UUID.randomUUID().toString()
+    def oauthParams = [
+        response_type: "code",
+        scope: "offline_access",
+        audience: "app.energycurb.com/api",
+        client_id: appSettings.clientId,
+        connection: "Users",
+        state: state.oauthInitState,
+        redirect_uri: callbackUrl
+    ]
+    redirect(location: "${curbLoginUrl}?${toQueryString(oauthParams)}")
+}
+
+def callback() {
+    def code = params.code
+    def oauthState = params.state
+    if (oauthState == state.oauthInitState) {
+        def tokenParams = [
+            grant_type: "authorization_code",
+            code: code,
+            client_id: appSettings.clientId,
+            client_secret: appSettings.clientSecret,
+            redirect_uri: callbackUrl
+        ]
+        httpPostJson([uri: curbTokenUrl, body: tokenParams]) {
+            resp ->
+                state.refreshToken = resp.data.refresh_token
+              state.authToken = resp.data.access_token
+
+            getCurbLocations()
+        }
+        if (state.authToken) {
+            success()
+        } else {
+            fail()
+        }
+    } else {
+        log.error "callback() failed oauthState != state.oauthInitState"
+    }
+}
+
+private removeChildDevices(delete) {
+    delete.each {
+        deleteChildDevice(it.deviceNetworkId)
+    }
+}
+
+def getCurbLocations() {
+    def params = [
+        uri: "http://app.energycurb.com",
+        path: "/api/locations",
+        headers: ["Authorization": "Bearer ${state.authToken}"]
+    ]
+
+    try {
+        httpGet(params) {
+            resp ->
+                def locationNameList = []
+            def locationLookup = []
+            resp.data.each {
+                locationNameList.push(it.name)
+                locationLookup.push(it)
+            }
+            state.locationNames = locationNameList
+            state.locationLookup = locationLookup
+        }
+    } catch (e) {
+        log.error "something went wrong: $e"
+    }
+}
+
+def updateSelectedLocationId() {
+  def location = ""
+    state.locationLookup.each {
+        if (it.name == settings.curbLocation) {
+            location = it.id
+        }
+    }
+    state.location = location
+}
+
+def updateChildDevice(dni, label, value) {
+    def device = getChildDevice(dni)
+    if (device == null) {
+      device = createChildDevice(dni, label)
+    }
+    device.handlePower(value)
+}
+
+def createChildDevice(dni, label) {
+    return addChildDevice("curb", "Curb Power Meter", dni, null, [name: "${dni}", label: "${label}"])
+}
+
+def getDevices() {
+    def params = [
+        uri: "https://app.energycurb.com",
+        path: "/api/latest/${state.location}",
+        headers: ["Authorization": "Bearer ${state.authToken}"],
+        requestContentType: 'application/json'
+    ]
+    asynchttp_v1.get(processDevices, params)
```

Comment:
If this fails, then devices won't get added but the SmartApp will get installed. This can lead to a bad experience, so it is suggested that this be converted to a synchronous call.